File: frontend/src/api/filters.ts
export interface IRange {
    min: number,
    max: number
}

export interface IFilters {
    ata_points_range: IRange,
    ata_score_range: IRange,
    passage_letters: string[],
    src_langs: string[],
    tgt_langs: string[],
    error_types: string[],
    severities: string[],
    ata_codes: string[]
}

export async function getFilters(): Promise<IFilters> {
    const response = await fetch('/api/filters', { method: 'POST' })
    if (!response.ok) return {
        ata_points_range: { min: 0, max: 0 },
        ata_score_range: { min: 0, max: 0 },
        passage_letters: [],
        src_langs: [],
        tgt_langs: [],
        error_types: [],
        severities: [],
        ata_codes: []
    }

    const json = (await response.json()) as IFilters
    return {
        ata_points_range: {
            min: parseInt(json?.ata_points_range?.min?.toString() ?? 0),
            max: parseInt(json?.ata_points_range?.max?.toString() ?? 0)
        },
        ata_score_range: {
            min: parseInt(json?.ata_score_range?.min?.toString() ?? 0),
            max: parseInt(json?.ata_score_range?.max?.toString() ?? 0),
        },
        passage_letters: json?.passage_letters?.sort() ?? [],
        src_langs: json?.src_langs?.sort() ?? [],
        tgt_langs: json?.tgt_langs?.sort() ?? [],
        error_types: json?.error_types?.sort() ?? [],
        severities: json?.severities?.sort() ?? [],
        ata_codes: json?.ata_codes?.sort() ?? []
    }
}

File: api/filters.py
#!/usr/bin/env python3

from json import dumps
from flask import (
    Blueprint
)

from ..db import ata_connection

bp = Blueprint('filters_api', __name__, url_prefix='/api')

@bp.route('/filters', methods=('POST',))
def get_counts():
    error = None

    try:
        error_list = ata_connection.fetch_error_list()
        if error is None:
            ret = {
                'ata_points_range': {
                    'min': str(error_list['ata_points'].min()),
                    'max': str(error_list['ata_points'].max())
                },
                'ata_score_range': {
                    'min': str(error_list['ata_score'].min()),
                    'max': str(error_list['ata_score'].max())
                },
                'passage_letters': error_list['text_letter'].unique().tolist(),
                'src_langs': error_list['src_lang'].unique().tolist(),
                'tgt_langs': error_list['tgt_lang'].unique().tolist(),
                'error_types': error_list['name'].unique().tolist(),
                'severities': error_list['severity'].unique().tolist(),
                'ata_codes': error_list['ata_code'].unique().tolist()
            }
            return dumps(ret)
    except Exception as e:
        error = e
        pass

    return {
        'error': error or "Counts could not be retrieved.",
    }
